GCE: Add Alpha feature "Network Tiers" for external L4 load balancers

yujuhong opened the pull request:
**Special notes for your reviewer**:
The PR has been manually tested in a GCE e2e cluster for the following conditions:
  1. When `network-tier` is not enabled in gce.conf, network tier annotations are completely ignored by the controller.
  2. When  `network-tier` is enabled in gce.conf:
     * Service w/ Standard tier: create a standard-tier LB.
     * Update Service to use a different tier: tear down the existing forwarding rule and release the IP before creating a new LB.
     * Service w/ an invalid tier value: `ensureExternalLoadBalancer()` returns an error, and controller emits an event.
     * Service w/ a user-owned static IP: check if the tier matches, if not, returns an error and emits an event.

I uploaded an e2e test #51483. You're welcome to review that one too.

**Release note**:
<!--  Steps to write your release note:
1. Use the release-note-* labels to set the release note state (if you have access)
2. Enter your extended release note in the below block; leaving it blank means using the PR title as the release note. If no release note is required, just write `NONE`.
-->
```release-note
GCE: Service object now supports "Network Tiers" as an Alpha feature via annotations.
```

yujuhong commented:
@nicksardo @bowei I split the changes into multiple commits so that it'd be easier to review. If you prefer multiple PRs, let me know.

nicksardo commented on pkg/cloudprovider/providers/gce/gce_addresses_fakes.go:
```diff
@@ -106,6 +106,18 @@ func (cas *FakeCloudAddressService) GetRegionAddress(name, region string) (*comp
+	if _, exists := cas.addrsByRegionAndName[region][name]; !exists {
+		return makeGoogleAPINotFoundError("")
+	}
+	delete(cas.addrsByRegionAndName[region], name)
```
`delete(cas.reservedAddrs, addr.Address)`

nicksardo commented on pkg/cloudprovider/providers/gce/gce_loadbalancer_external.go:
```diff
@@ -913,3 +978,66 @@ func ensureStaticIP(s CloudAddressService, name, serviceName, region, existingIP
+	if isNotFound(err) {
+		return nil
+	}
+	if err != nil {
```
nit: else if

nicksardo commented on pkg/cloudprovider/providers/gce/gce_loadbalancer_external.go:
```diff
@@ -913,3 +978,66 @@ func ensureStaticIP(s CloudAddressService, name, serviceName, region, existingIP
+
+// reconcileFwdRuleNetworkTiers check the network tiers of existing forwarding
+// rule and delete the rule if the tier does not matched the desired tier.
+func reconcileFwdRuleNetworkTiers(s CloudForwardingRuleService, region, name, lbRef string, desiredNetTier NetworkTier) error {
```
Using the word `reconcile` seems strange to me here because you're only deleting the forwarding rule/address if the tier needs changed. 

I suggest wripping the two `reconcile` funcs apart and introduce a higher level func which checks tier equality and deletes the resources if necessary.

```go
func deleteWrongNetworkTieredResources(...){ 
  fwdRuleOk, err := fwdRuleNetworkTierEquals(gce, region, name, lbRef, desiredNetTier)
  if err!= nil {
...
} 
if !fwdRuleOk {
  err = s.DeleteRegionForwardingRule(name, region)
}

 addrOk, err := addressNetworkTierEquals(gce, region, name, lbRef, desiredNetTier)
  if err!= nil {
...
} 
if !addrOk {
  err = s.DeleteRegionAddress(name, region)
}

}

yujuhong replied:
Agree that the names are confusing. The function was initially intended to do more, but that didn't pan out.
Renamed the functions. I still keep the address and forwarding rule verification in two functions to make testing easier.

yujuhong reviewed (commented):
Addressed comments. Thanks!

nicksardo commented on pkg/cloudprovider/providers/gce/gce_loadbalancer_external.go:
```diff
@@ -883,27 +925,46 @@ func (gce *GCECloud) firewallObject(name, region, desc string, sourceRanges nets
+			addressObj.Address = existingIP
+		}
+		if err = s.ReserveRegionAddress(addressObj, region); err != nil {
+			if !isHTTPErrorCode(err, http.StatusConflict) {
```
Or BadRequest

nicksardo commented on pkg/cloudprovider/providers/gce/gce_loadbalancer_external.go:
```diff
@@ -883,27 +925,46 @@ func (gce *GCECloud) firewallObject(name, region, desc string, sourceRanges nets
+			addressObj.Address = existingIP
+		}
+		if err = s.ReserveAlphaRegionAddress(addressObj, region); err != nil {
+			if !isHTTPErrorCode(err, http.StatusConflict) {
```
Or BadRequest

nicksardo commented:
I don't see anything wrong with it, though we should get another pair of eyes. @bowei @MrHohn ?

yujuhong commented:
Flakes caused by #51379

/retest

yujuhong commented:
/test pull-kubernetes-e2e-gce-bazel

MrHohn commented on pkg/cloudprovider/providers/gce/gce_loadbalancer_external.go:
```diff
@@ -72,6 +69,21 @@ func (gce *GCECloud) ensureExternalLoadBalancer(clusterName, clusterID string, a
+	// Check the current and the desired network tiers. If they do not match,
+	// tear down the existing resources -- delete the forwarding rule, and
+	// release the IP.
+	netTier, err := gce.getServiceNetworkTier(apiService)
```
Could we also gate network tier outside of `getServiceNetworkTier()`? Like `reconcileXXXNetworkTiers()` below, so that we don't trigger the logic at all.

yujuhong replied:
I did that intentionally since `getServiceNetworkTier()` should always return the default tier if the feature is disabled. If I could, I'd actually prefer gating it only at the gce API calls so that we can reuse the same flow when the feature graduates to Beta. Otherwise, the beta code will be completely new code which needs to be tested all over again. I think the current implementation provides sufficient "gating" but feel free to disagree :-)

MrHohn replied:
Humm...I think I some how mixed the concepts of gating alpha/beta GCE API and gating the actual feature. IMHO:
- Gate alpha/beta API means no calls to them.
- Gate the feature means no triggers to the logic.

Given that the GCE provider gate is about gating API but not feature, I think it makes sense to do it this way...

MrHohn replied:
> I think the current implementation provides sufficient "gating" but feel free to disagree :-)

But yeah technically this is sufficient :)

MrHohn commented on pkg/cloudprovider/providers/gce/gce_loadbalancer_external.go:
```diff
@@ -72,6 +69,21 @@ func (gce *GCECloud) ensureExternalLoadBalancer(clusterName, clusterID string, a
+		glog.Errorf("EnsureLoadBalancer(%s): failed to get the desired network tier: %v", lbRefStr, err)
+		return nil, err
+	}
+	glog.V(2).Infof("EnsureLoadBalancer(%s): desired network tier %q ", lbRefStr, netTier)
```
User that doesn't enable network tier may be seeing this "network tier" log and they may think it is mistakenly enabled?

yujuhong replied:
Bumped it to V(4) so that it doesn't show up in GKE clusters.

MrHohn replied:
Sorry for switching side...Actually I'd prefer to keep this in V(2) for better debuggability in case things go wrong in production. (I'd like to feature gate it, or maybe it just doesn't matter)

MrHohn commented on pkg/cloudprovider/providers/gce/gce_loadbalancer_external.go:
```diff
@@ -991,15 +987,28 @@ func (gce *GCECloud) getServiceNetworkTier(svc *v1.Service) (NetworkTier, error)
 
-// reconcileFwdRuleNetworkTiers check the network tiers of existing forwarding
+func (gce *GCECloud) deleteWrongNetworkTieredResources(lbName, lbRef string, desiredNetTier NetworkTier) error {
+	if !gce.AlphaFeatureGate.Enabled(AlphaFeatureNetworkTiers) {
```
Humm..Seems like the feature gate check is moved inside functions. I think having it outiside may make it clear about what are involved.

yujuhong replied:
Restored it.

MrHohn reviewed (commented):
Looks good and logic makes sense to me.

Couple minor comments.

yujuhong commented on pkg/cloudprovider/providers/gce/gce_annotations.go:
```diff
@@ -39,6 +44,17 @@ const (
+	// NetworkTierAnnotationKey is annotated on a Service object to indicate which
+	// network tier a GCP LB should use. The valid values are "Standard" and
+	// "Premium" (default).
+	NetworkTierAnnotationKey      = "cloud.google.com/network-tier"
```
I can't find it but I think it's recommended not to use `alpha` in the annotation anymore.

nicksardo replied:
Good question. I suggest double checking with thockin.

yujuhong replied:
/cc @thockin

yujuhong replied:
Spoke to @thockin offline. This is ok.

yujuhong commented on pkg/cloudprovider/providers/gce/gce_loadbalancer_external.go:
```diff
@@ -282,8 +286,8 @@ func (gce *GCECloud) ensureExternalLoadBalancer(clusterName, clusterID string, a
-		glog.Infof("EnsureLoadBalancer(%v(%v)): creating forwarding rule, IP %s", loadBalancerName, serviceName, ipAddressToUse)
-		if err := gce.createForwardingRule(loadBalancerName, serviceName.String(), gce.region, ipAddressToUse, ports); err != nil {
+		glog.Infof("EnsureLoadBalancer(%v(%v)): creating forwarding rule, IP %s (tier: %s)", loadBalancerName, serviceName, ipAddressToUse, netTier)
+		if err := createForwardingRule(gce, loadBalancerName, serviceName.String(), gce.region, ipAddressToUse, gce.targetPoolURL(loadBalancerName), ports, netTier); err != nil {
```
Yep. I could fix the targetPool code too with more mocking, but I'd prefer leave them alone for this change.

yujuhong commented on pkg/cloudprovider/providers/gce/gce_loadbalancer_external.go:
```diff
@@ -883,27 +906,46 @@ func (gce *GCECloud) firewallObject(name, region, desc string, sourceRanges nets
+			addressObj.Address = existingIP
+		}
+		if err = s.ReserveAlphaRegionAddress(addressObj, region); err != nil {
+			if !isHTTPErrorCode(err, http.StatusConflict) {
```
I like checking the err code close to the call site, but sure I'll change it.

yujuhong replied:
Done.

yujuhong commented on pkg/cloudprovider/providers/gce/gce_loadbalancer_naming.go:
```diff
@@ -84,6 +84,11 @@ func makeBackendServiceDescription(nm types.NamespacedName, shared bool) string
 // External Load Balancer
 
+// makeServiceDescription is used to generate descriptions for forwarding rules and addresses.
+func makeServiceDescription(serviceName string) string {
```
Can we punt this to when we have more than two functions? :-)

yujuhong commented on pkg/cloudprovider/providers/gce/gce_loadbalancer_external.go:
```diff
@@ -446,7 +446,19 @@ func verifyUserRequestedIP(s CloudAddressService, region, requestedIP, fwdRuleIP
+		}
+		netTier := NetworkTierGCEValueToType(netTierStr)
+		if netTier != desiredNetTier {
+			glog.Errorf("verifyUserRequestedIP: requested static IP %q (name: %s) for LB %s has network tier %s, need %s.", requestedIP, existingAddress.Name, lbRef, netTier, desiredNetTier)
```
Error will be included in the k8s even and probably should not have too many function-specific information.

yujuhong commented on pkg/cloudprovider/providers/gce/gce_loadbalancer_external_test.go:
```diff
@@ -85,3 +122,54 @@ func TestVerifyRequestedIP(t *testing.T) {
 	}
 }
+
+func TestCreateForwardingRule(t *testing.T) {
```
Done.

yujuhong commented:
@MrHohn the new commit addressed (most) of your comments.

nicksardo commented:
Since you're making the BadRequest/StatusConflict fix in this PR, could you also change the fake service to return the both types.

```go
func (cas *FakeCloudAddressService) ReserveAlphaRegionAddress(addr *computealpha.Address, region string) error {
	if addr.Address == "" {
		addr.Address = fmt.Sprintf("1.2.3.%d", cas.count)
		cas.count++
	}

	if addr.AddressType == "" {
		addr.AddressType = string(schemeExternal)
	}

	if cas.reservedAddrs[addr.Address] {
		msg := "IP in use"
		// Depending on the type of address, different error messages may be returned.
		switch lbScheme(addr.AddressType) {
		case schemeExternal:
			return makeGoogleAPIError(http.StatusBadRequest, msg)
		default:
			return makeGoogleAPIError(http.StatusConflict, msg)
		}
	}

	if _, exists := cas.addrsByRegionAndName[region]; !exists {
		cas.addrsByRegionAndName[region] = make(map[string]*computealpha.Address)
	}

	if _, exists := cas.addrsByRegionAndName[region][addr.Name]; exists {
		return makeGoogleAPIError(http.StatusConflict, "name in use")
	}

	cas.addrsByRegionAndName[region][addr.Name] = addr
	cas.reservedAddrs[addr.Address] = true
	return nil
}
```

bowei commented on pkg/cloudprovider/providers/gce/gce_forwardingrule.go:
```diff
@@ -133,3 +133,15 @@ func (gce *GCECloud) DeleteRegionForwardingRule(name, region string) error {
 	return gce.waitForRegionOp(op, region, mc)
 }
+
+// TODO(yujuhong): retire this function once Network Tiers becomes Beta in GCP.
```
link to the same issue as above

yujuhong replied:
Done

bowei commented on pkg/cloudprovider/providers/gce/gce_loadbalancer_external.go:
```diff
@@ -426,7 +439,7 @@ func (gce *GCECloud) DeleteExternalTargetPoolAndChecks(name, region, clusterID s
 // verifyUserRequestedIP checks the user-provided IP  to see whether it can be
 // used for the LB.  It also returns whether the IP is considered owned by the
```
Can you clarify this comment as the function can now return (true, err != nil), seems confusing.

yujuhong replied:
Good point. I changed the function to never return (true, non-nil-err). If the err is non-nil, the boolean value should not be used anyway.

bowei commented on pkg/cloudprovider/providers/gce/gce_loadbalancer_external.go:
```diff
@@ -913,3 +974,73 @@ func ensureStaticIP(s CloudAddressService, name, serviceName, region, existingIP
+	if existingTier == desiredNetTier {
+		return nil
+	}
+	glog.V(2).Infof("%s: Network tiers do not match; existing forwarding rule: %q, desired: %q", logPrefix, existingTier, desiredNetTier)
```
probably just combine into one log

yujuhong replied:
Done

bowei commented on pkg/cloudprovider/providers/gce/gce_loadbalancer_external.go:
```diff
@@ -913,3 +974,73 @@ func ensureStaticIP(s CloudAddressService, name, serviceName, region, existingIP
+	if existingTier == desiredNetTier {
+		return nil
+	}
+	glog.V(2).Infof("%s: Network tiers do not match; existing address: %q, desired: %q", logPrefix, existingTier, desiredNetTier)
```
combine these log lines...

yujuhong replied:
Done

bowei commented on pkg/cloudprovider/providers/gce/gce_util.go:
```diff
@@ -153,3 +153,20 @@ func isNotFoundOrInUse(err error) bool {
+	return isHTTPErrorCode(err, http.StatusForbidden)
+}
+
+func handleNetworkTierGetError(err error) (string, error) {
```
is this logic not possible to just put in the GetAlphaRegionAddress, ... functions?

this is a super generic function name...

yujuhong replied:
It's shared by two files -- gce_addresses.go and gce_forwardingrule.go.
I renamed it to `handleAlphaNetworkTierGetError()` to make it slightly more explicit, and also added a TODO with a bug number to remove this.

yujuhong commented on pkg/cloudprovider/providers/gce/gce_addresses.go:
```diff
@@ -132,3 +132,15 @@ func (gce *GCECloud) GetRegionAddressByIP(region, ipAddress string) (*compute.Ad
 	return nil, makeGoogleAPINotFoundError(fmt.Sprintf("Address with IP %q was not found in region %q", ipAddress, region))
 }
+
+// TODO(yujuhong): retire this function once Network Tiers becomes Beta in GCP.
```
Done

yujuhong commented on pkg/cloudprovider/providers/gce/gce_annotations.go:
```diff
@@ -39,6 +44,17 @@ const (
+	NetworkTierAnnotationStandard = "Standard"
+	NetworkTierAnnotationPremium  = "Premium"
+
+	NetworkTierStandard NetworkTier = NetworkTierAnnotationStandard
```
I prefer keeping the annotation key/values pure strings (not an internal `NetworkTier` type) since it's essentially our API.

yujuhong commented on pkg/cloudprovider/providers/gce/gce_annotations.go:
```diff
@@ -77,3 +93,42 @@ func GetLoadBalancerAnnotationBackendShare(service *v1.Service) bool {
+// GetServiceNetworkTier returns the network tier of GCP load balancer
+// which should be assembled, and an error if the specified tier is not
+// supported.
+func GetServiceNetworkTier(service *v1.Service) (NetworkTier, error) {
```
Any suggestion? I am bad with naming and the `Get` prefix makes the intent of the function clear to me.

yujuhong commented:
@bowei addressed comments in the new commit and rebased.

> Since you're making the BadRequest/StatusConflict fix in this PR, could you also change the fake service to return the both types.

@nicksardo, added a commit for that. Will rebase if your PR merges first.

yujuhong commented:
Also filed #51668 for an existing bug in external LB that this PR doesn't fix.

yujuhong commented:
@bowei PTAL, thanks!

bowei commented:
/lgtm

yujuhong committed:
GCE: Add "Network Tiers" as an Alpha feature for L4 load balancers

This feature supports specifying what network tier (premium, standard)
to use for the load balancer backing the Service (type=LoadBalancer).

yujuhong commented on pkg/cloudprovider/providers/gce/gce_addresses_fakes.go:
```diff
@@ -64,16 +63,27 @@ func (cas *FakeCloudAddressService) ReserveAlphaRegionAddress(addr *computealpha
 	if cas.reservedAddrs[addr.Address] {
-		return &googleapi.Error{Code: http.StatusConflict}
+		msg := "IP in use"
+		// Depending on the type of address, different error messages may be returned.
```
Added.

yujuhong commented on pkg/cloudprovider/providers/gce/gce_alpha.go:
```diff
@@ -22,15 +22,22 @@ import (
+	//
+	// Allows Services backed by a GCP load balancer to choose what network
+	// tier to use. Currently supports "Standard" and "Premium" (default).
+	AlphaFeatureNetworkTiers = "network-tiers"
```
Done.

yujuhong commented:
Rebased and addressed the nits. Re-applying lgtm based on https://github.com/kubernetes/kubernetes/pull/51301#issuecomment-326457931
Will probably need to rebase at some point though.

yujuhong commented:
Rebased again and verified that the rebased PR still passed the e2e test.

yujuhong commented:
> Rebased again and verified that the rebased PR still passed the e2e test.

Re-applied the lgtm.

yujuhong commented:
/test pull-kubernetes-kubemark-e2e-gce

kubemark issue has already been fixed by a revert. Re-testing

yujuhong commented:
/test pull-kubernetes-kubemark-e2e-gce

yujuhong commented:
FWIW, the PR has already passed the kubemark test: https://k8s-gubernator.appspot.com/build/kubernetes-jenkins/pr-logs/pull/51301/pull-kubernetes-kubemark-e2e-gce/49968/

krzyzacy commented:
@spxtr ^^ is this mean something wrong when reporting back (crier)? any idea how to recover?

spxtr commented:
/test pull-kubernetes-kubemark-e2e-gce

krzyzacy commented:
http://prow.k8s.io/?type=presubmit&job=pull-kubernetes-kubemark-e2e-gce

I think the job is just backlogged... the test was from a different trigger earlier and you triggered it again and now it's queueing again

spxtr commented:
Looks like this job is regularly taking 3+ hours to complete, including queue time. Welp.

yujuhong commented:
> I think the job is just backlogged... the test was from a different trigger earlier and you triggered it again and now it's queueing again

Thanks for checking. I'll check the queue next time before triggering....

yujuhong commented:
Failed due to some infra flake: https://k8s-gubernator.appspot.com/build/kubernetes-jenkins/pr-logs/pull/51301/pull-kubernetes-kubemark-e2e-gce/50032

/test pull-kubernetes-kubemark-e2e-gce

cblecker commented:
/test pull-kubernetes-kubemark-e2e-gce
